### PR TITLE
Issue 36101: Category selection tree in edit product is not in the sa…

### LINKED ISSUE #36101: Category selection tree in edit product is not in the same order as set in Catalog->Categories
### Preconditions and environment

Development branch and latest release (as well as earlier ones).  Verified on latest development branch.

### Steps to reproduce

1. Create a new category under Catalog->Categories
2. Move the new category to the top of the tree
3. Edit a product and click on the Categories select... drop down

### Expected result

- New category should be at the top of the tree as per display order set in Catalog->Categories

### Actual result

The new category is at the bottom of the tree - suggesting the category tree under edit product is in creation order

Makes administering  products difficult in large catalogues more difficult

### Additional information

Current behaviour is not what I would say was intuitive or expected.

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by rogerdz**
…me order as set in Catalog->Categories

### Description (*)
Category selection tree in edit product is not in the same order as set in Catalog->Categories

### Fixed Issues (if relevant)

1. Fixes magento/magento2#36101

### Manual testing scenarios (*)

1. Create a new category under Catalog->Categories
2. Move the new category to the top of the tree
3. Edit a product and click on the Categories select... drop down

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by rogerdz** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php [hunk 1]
```diff
@@ -442,6 +444,8 @@ private function retrieveCategoriesTree(int $storeId, array $shownCategoriesIds)
 
         $collection->addAttributeToFilter('entity_id', ['in' => array_keys($shownCategoriesIds)])
             ->addAttributeToSelect(['name', 'is_active', 'parent_id'])
+            ->addAttributeToSort('level', 'asc')
```
Comment (edenduong) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php]: 'asc' should be changed to ```SORT_ORDER_ASC``` constant

Comment (edenduong) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php]: 'level'  and 'position' has the constant in \Magento\Catalog\Api\Data\CatalogInterface
    const KEY_POSITION = 'position';
    const KEY_LEVEL = 'level';

**Review by edenduong — CHANGES_REQUESTED** [review]
Thank you for your contribution. Please check my comment

**Comment by rogerdz** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php (hunk 1 above)
Comment (rogerdz) [reply app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php]: thank @edenduong , I pushed a new commitI

**Review by edenduong — CHANGES_REQUESTED** [review]
Could you please check and modify the Unit Test? It is failed because of the new logic. 
Thanks you.

**Comment by rogerdz** [comment]
> 

@edenduong I fixed the unit test.
@magento run all tests

**Review by edenduong — APPROVED** [review]
Thank you for your contribution !

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php [hunk 2]
```diff
@@ -52,6 +54,7 @@ class Categories extends AbstractModifier
      * @var array
      * @deprecated 101.0.0
      * @since 101.0.0
+     * @see Not use anymore
```
Comment (Den4ik) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php]: `@see` used for linking to the class|function etc. In this case I prefer approve failed static test as legacy without changes like that

@sidolov what's your opinion?

**Review by Den4ik — CHANGES_REQUESTED** [review]
@rogerdz Thanks for work. Please look at my comment

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php (hunk 2 above)
Comment (rogerdz) [reply app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php]: Can i use @codingStandardsIgnoreLine to ignore it?

Comment (rogerdz) [reply app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php]: @Den4ik @sidolov  please advise

Comment (rogerdz) [reply app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php]: I fixed

**Comment by rogerdz** [comment]
@magento run all tests

**Review by edenduong — APPROVED** [review]

**Comment by engcom-Lima** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento `2.4-develop`

**Manual testing scenario:**

1. Create a new category under Catalog->Categories
2. Move the new category to the top of the tree
3. Edit a product and click on the Categories select... drop down

**Before:** :heavy_multiplication_x: Category was not displaying in correct order during product edit.
**Category creation page:**
Screenshot 2022-12-20 at 12 30 30 PM

**Product edit page:**
Screenshot 2022-12-20 at 12 29 55 PM


**After:** :heavy_check_mark:  Now it is working fine.

**Category creation page:**
Screenshot 2022-12-20 at 1 05 51 PM

**Product edit page:**
Screenshot 2022-12-20 at 1 06 29 PM

Several builds are failed. Hence, moving this PR to extended testing.

**Comment by engcom-Charlie** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Charlie** [comment]
It seems bot moved it from Extended Testing to Testing hence moving it back.

**Comment by engcom-Charlie** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento give me test instance

**Comment by engcom-Charlie** [comment]
@magento run Static Tests, Unit Tests, Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
Hi @rogerdz,

Thank you for your contribution!

In order to move further on this PR, we need green build. Can you please look into below error from Unit test failures. 
```
1) Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\CategoriesTest::testModifyMetaLocked with data set #0 (true)
Error: Call to a member function addAttributeToSort() on null

/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:450
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:361
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:286
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:154
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Test/Unit/Ui/DataProvider/Product/Form/Modifier/CategoriesTest.php:217

2) Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\CategoriesTest::testModifyMetaLocked with data set #1 (false)
Error: Call to a member function addAttributeToSort() on null

/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:450
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:361
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:286
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Categories.php:154
/opt/homebrew/var/www/PR_36368/app/code/Magento/Catalog/Test/Unit/Ui/DataProvider/Product/Form/Modifier/CategoriesTest.php:217
```

**Comment by rogerdz** [comment]
@magento run all tests

**Comment by rogerdz** [comment]
Hi @engcom-Charlie 
I fixed Unit test

**Comment by engcom-Charlie** [comment]
> Hi @engcom-Charlie I fixed Unit test

Thank you @rogerdz!

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by engcom-Charlie** [comment]
Few test failures in recent 2 builds for functional B2B, CE and EE are not consistent and remaining are known issues. Hence moving it to Merge in Progress.

**Functional B2B: Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/36104/599123ffb60f1c6f1a3ed3d92adfec09/Functional/allure-report-b2b/index.html
image

**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/36104/03724af21b0357de0b8b1786ecb61941/Functional/allure-report-b2b/index.html
image


**Functional CE: Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/36104/d9fe750e5f973342decbab2c2a4c23cb/Functional/allure-report-ce/index.html#categories
image


**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/36104/c8d4e7a7e7f2afe44454eaf0bb25f75a/Functional/allure-report-ce/index.html#categories
image


**Functional EE: Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/36104/147c4818408780de77c4eb68766850b9/Functional/allure-report-ee/index.html
image


**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/36104/6b6df46e7e1044d3c4f0b88ec2623a16/Functional/allure-report-ee/index.html
image

**Known issues in CE, EE, B2B:**
1. StorefrontAddBundleDynamicProductToShoppingCartWithDisableMiniCartSidebarTest: [#ACQE-6786](https://jira.corp.adobe.com/browse/ACQE-6786)
4. StorefrontCashOnDeliveryPaymentForSpecificCountryTest: [#ACQE-6774](https://jira.corp.adobe.com/browse/ACQE-6774)
3. StorefrontCreateOrderAllQuantityGroupedProductOptionDefaultStockTest: [#ACQE-6331](https://jira.corp.adobe.com/browse/ACQE-6331)
4. StorefrontLoginFormCheckDuplicateValidateMessageTest: [#ACQE-6799](https://jira.corp.adobe.com/browse/ACQE-6799)
5. TierPricingWhenPriceScopeIsWebsiteWorkingProperlyWithMultipleCurrenciesConfiguredTest: [#ACQE-6695](https://jira.corp.adobe.com/browse/ACQE-6695)

**Pull request merged**